cli/src/trade/trade.tsx

```diff
@@ -250,6 +269,180 @@ export const LegMaker = ({
     [buildLeg, setStage, onLegConfirm, addMessage, chosenPairInfo, positionSize]
   )
 
+  const cli = useCli()
+  const pool = useOption('pool')
+  const asset = useOption('asset')
+  const trade = useOption('trade')
+  const sp = useOption('sp')
+  const priceRange = useOption('range')
+  const amount = useOption('amount')
+  const { pairs } = usePools()
+
+  useEffect(() => {
+    if (!cli || !pool || !pairs) {
+      return
+    }
+
+    if (stage !== TradeStage.PoolSelection) {
+      return
+    }
+
+    const pair = pairs.find((p) =>
+      [
+        `${p.token0}/${p.token1}`.toLowerCase(),
+        `${p.token1}/${p.token0}`.toLowerCase()
+      ].includes(pool.toLowerCase())
+    )
+
+    if (pair === undefined) {
+      addMessage(`Pool not found: ${pool.toUpperCase()}`, { color: 'red' })
+      setStage(TradeStage.Empty)
+      return
+    }
+
+    onPoolSelected(pair)
+  }, [cli, addMessage, pairs, pool, stage, setStage, onPoolSelected])
+
+  useEffect(() => {
+    if (
+      !cli ||
+      !asset ||
+      !chosenPairInfo.ready ||
+      stage !== TradeStage.QuoteAsset
+    ) {
+      return
+    }
+
+    let choice = 0
+
+    if (chosenPairInfo.c0Info.symbol.toLowerCase() === asset.toLowerCase()) {
+      choice = 1
+    } else if (
+      chosenPairInfo.c1Info.symbol.toLowerCase() === asset.toLowerCase()
+    ) {
+      choice = 2
+    }
+
+    if (choice === 0) {
+      addMessage(`Invalid asset: ${asset.toUpperCase()}`, { color: 'red' })
+      setStage(TradeStage.Empty)
+      return
+    }
+
+    onQuoteAssetSubmit(choice)
+  }, [
+    addMessage,
+    asset,
+    chosenPairInfo,
+    cli,
+    onQuoteAssetSubmit,
+    setStage,
+    stage
+  ])
+
+  useEffect(() => {
+    if (!cli || stage !== TradeStage.PutCall) {
+      return
+    }
+
+    let choice = 0
+
+    if (trade?.toLowerCase() === 'put') {
+      choice = 1
+    } else if (trade?.toLowerCase() === 'call') {
+      choice = 2
+    }
+
+    if (choice === 0) {
+      addMessage(
+        `Invalid trade (should be PUT or CALL): ${trade?.toUpperCase()}`,
+        { color: 'red' }
+      )
+      setStage(TradeStage.Empty)
+      return
+    }
+
+    onPutCallSelected(choice)
+  }, [addMessage, cli, onPutCallSelected, stage, setStage, trade])
+
+  useEffect(() => {
+    if (!cli || stage !== TradeStage.StrikeAmount) {
+      return
+    }
+
+    if (isNaN(Number(sp))) {
+      addMessage(`Invalid striker price: ${sp}`, { color: 'red' })
+      setStage(TradeStage.Empty)
+      return
+    }
+
+    onStrikeAmountSubmit(Number(sp))
+  }, [addMessage, cli, stage, sp, setStage, onStrikeAmountSubmit])
+
+  useEffect(() => {
+    if (!cli || stage !== TradeStage.Width) {
+      return
+    }
+
+    if (!onWidthSubmit(String(priceRange))) {
+      setStage(TradeStage.Empty)
+    }
+  }, [addMessage, cli, stage, priceRange, setStage, onWidthSubmit])
+
+  const stageRef = useRef<TradeStage>(TradeStage.Quantity)
```

Comment:
Why is there a separate ref variable for stage?

Reply:
Stage alone is not sufficient to specify the step. It keeps falling into the useEffect of Quantity stage without this ref.